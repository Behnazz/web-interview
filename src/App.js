import React, { Component } from 'react'
import Buttons from './Components/Buttons'
import axios from 'axios'
import moment from 'moment'
import logo from './logo.png'
import { API_ENDPOINT } from './config'
import UserSection from './Components/UserSection'
import './App.scss'

const consultant = ['gp', 'therapist', 'physio', 'specialist']
const captalizeConsultantTitle = consultantArray =>
  consultantArray.map(name =>
    name
      .split(' ')
      .map(word => word[0].toUpperCase() + word.slice(1).toLowerCase())
      .join('')
  )

const consultantTitle = captalizeConsultantTitle(consultant)

const appTypeMedium = ['Audio', 'Video']

class App extends Component {
  constructor(props) {
    super(props)

    this.state = {
      selectedAppointmentType: '',
      selectedConsultantType: null,
      availableSlots: [],
      availableSlotsByConsultantType: [],
      selectedDate: '',
      symptoms: '',
      userInfo: {},
      error: null,
      loading: false,
    }
  }

  async componentDidMount() {
    this.getData()
    this.getUser()
  }

  getUser = async () => {
    this.setState({
      loading: true,
    })
    try {
      const userData = await axios.get(`${API_ENDPOINT}/users/1`)

      this.setState({ userInfo: userData.data, loading: false })
    } catch (error) {
      this.setState({
        error,
        loading: false,
      })
    }
  }
  getData = async () => {
    this.setState({ loading: true })
    try {
      const availableSlot = await axios.get(`${API_ENDPOINT}/availableSlots`)
      this.setState({ availableSlots: availableSlot.data, loading: false })
    } catch (error) {
      this.setState({
        error,
        loading: false,
      })
    }
  }

  handleConsultantClick = btnValue => {
    const filteredData = this.state.availableSlots.filter(item =>
      item.consultantType.includes(btnValue.toLowerCase())
    )
    this.setState({
      availableSlotsByConsultantType: filteredData,
      selectedConsultantType: btnValue,
    })
  }

  handleAppointmentClick = date => {
    this.setState({
      selectedDate: date,
    })
  }

  handleAppMedium = item => {
    this.setState({
      selectedAppointmentType: item,
    })
  }

  handleNote = e => {
    this.setState({
      symptoms: e.target.value,
    })
  }

  handleSubmit = async e => {
    e.preventDefault()
    const {
      symptoms,
      userInfo,
      selectedConsultantType,
      selectedAppointmentType,
      selectedDate,
      availableSlots,
    } = this.state
    const date = availableSlots.find(
      item => moment(item.time).format('MMM DD, h:mm a') === selectedDate
    )
    console.log(date)
    if (
      symptoms === '' ||
      userInfo === '' ||
      selectedAppointmentType === '' ||
      selectedConsultantType === '' ||
      selectedDate === ''
    ) {
      this.setState({
        error: 'Please fill the form completely',
      })
    } else {
      const body = {
        notes: symptoms,
        userId: userInfo.id,
        consultantType: selectedConsultantType,
        appointmentType: selectedAppointmentType,
        dateTime: date.time,
      }

      try {
        await axios.post('http://localhost:3010/appointments', body)
      } catch (error) {
        this.setState({
          error,
        })
      }
      this.setState({
        selectedAppointmentType: '',
        selectedConsultantType: null,
        selectedDate: '',
        symptoms: '',
        error: null,
      })
    }
  }

  render() {
    const {
      availableSlots,
      availableSlotsByConsultantType,
      selectedConsultantType,
      selectedAppointmentType,
      loading,
      userInfo,
      selectedDate,
      symptoms,
      error,
    } = this.state

    let slots =
      selectedConsultantType !== null
        ? availableSlotsByConsultantType
        : availableSlots
    return (
      <div className="app">
        <div className="app-header">
          <img
            src={logo}
            className="app-logo"
            alt="Babylon Health"
            data-testid="babylon-logo"
          />
        </div>

        <h1 className="h1">New appointment</h1>
        <UserSection
          firstName={userInfo.firstName}
          lastName={userInfo.lastName}
          avatar={userInfo.avatar}
        />
        <div className="section">
          <span className="icon">
            <i className="fa fa-stethoscope"></i>
            <h2>Consultant Type</h2>
          </span>
          <div className="container">
            {consultantTitle.map((item, index) => (
              <Buttons
                handleClick={this.handleConsultantClick}
                key={index}
                title={item}
                selectedButton={selectedConsultantType}
              />
            ))}
          </div>
          <div className="section">
            <span className="icon">
              <i className="fa fa-clock-o"></i>
              <h2>Appointment Date and time</h2>
            </span>
            <div className="container" data-testid="slot">
              {loading ? (
                <h3>loading</h3>
              ) : (
                slots.map(slot => (
                  <Buttons
                    title={moment(slot.time).format('MMM DD, h:mm a')}
                    handleClick={this.handleAppointmentClick}
                    key={slot.id}
                    selectedButton={selectedDate}
                  />
                ))
              )}
            </div>
          </div>
          <div className="section">
            <span className="icon">
              <i className="fa fa-video-camera"></i>
              <h2>Appointment Type</h2>
            </span>
            <div className="container">
              {appTypeMedium.map((item, index) => (
                <Buttons
                  handleClick={this.handleAppMedium}
                  key={index}
                  title={item}
                  selectedButton={selectedAppointmentType}
                />
              ))}
            </div>
          </div>
          <div className="section">
            {' '}
            <span className="icon">
              <i className="fa fa-comments-o"></i>
              <h2>Notes</h2>
            </span>
            <textarea
              aria-label="note for the symptoms"
              className="note"
              data-testid="symptoms"
              placeholder="Describe your symptoms"
              type="text"
              value={symptoms}
              onChange={this.handleNote}
            />
          </div>
          <hr />
          <div>
            {error && <h3 className="fillFormAlert">{error}</h3>}
            <button
              data-testid="button-xl"
              className="button-xl"
              onClick={this.handleSubmit}
            >
              Book
            </button>
          </div>
        </div>
      </div>
    )
  }
}

export default App
