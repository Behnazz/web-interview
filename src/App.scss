@import './constants.scss';

@font-face {
  font-family: Lato-Bold;
  font-family: Lato-Regular;
  src: url(./assets/fonts/Lato-Bold.ttf);
  src: url(./assets/fonts/Lato-Regular.ttf);
}

body {
  max-width: 900px;
  margin: 24px auto;
}
.app-user {
  display: flex;
  justify-content: flex-start;
  border-bottom: $border-standard;
}
.user-name {
  padding: 20px;
  font-family: 'Lato-Bold', sans-serif;
}

.app-user-logo {
  width: 80px;
  height: 80px;
}
.app-header {
  padding: $gutter-size--large;
  display: flex;
  justify-content: flex-end;
  img {
    max-height: 33px;
  }

  border-bottom: $border-standard;
}
.note {
  width: 100%;
  height: 150px;
  display: flex;
  justify-content: center;
  margin-bottom: 100px;
  border-radius: 10px;
  outline: 0;
  font-family: 'Lato-Bold', sans-serif;
  border: 1px solid grey;
}

.button-xl {
  box-sizing: border-box;
  padding: 20px;
  background-color: $primary-dark;
  color: whitesmoke;
  width: 100%;
  font-size: 1.5em;
  margin-right: 12px;
  border-radius: 10px;
  display: flex;
  justify-content: center;
  outline: 0;
  font-family: 'Lato-Bold', sans-serif;
  margin-top: 30px;
}

h2 {
  &.h6 {
    font-size: 24px;
    display: flex;
    margin-left: 10px;
    font-family: 'Lato-Bold', sans-serif;
  }
}

h1 {
  color: #2f333f;
  font-family: 'Lato-Bold', sans-serif;
}
.container {
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
}
.button {
  background-color: white;
  color: grey;
  outline: 0;
  box-sizing: border-box;
  border: solid 1px $primary-dark;
  border-radius: 30px;
  font-family: 'Lato-Regular', sans-serif;
  margin: 12px;
  padding: 10px 20px;
  transition: 0.3s;
}

.selected {
  background-color: $primary-dark;
  color: white;
}

.button:hover {
  background-color: $primary-dark;
  color: whitesmoke;
}

.icon {
  display: flex;
  align-items: center;
}

.fa {
  font-size: 28px;
  color: grey;
}

textarea::-webkit-input-placeholder {
  padding: 15px;
}

.section {
  margin-top: 20px;
}
@media only screen and (max-width: 600px) {
  body {
    max-width: 300px;
    margin: 24px auto;
  }
  .container {
    flex-direction: column;
    justify-content: center;
  }
}
