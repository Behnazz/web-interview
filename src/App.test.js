import React from 'react'
import ReactDOM from 'react-dom'
import {
  render,
  cleanup,
  fireEvent,
  waitForElement,
} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import axiosMock from 'axios'
import App from './App'

afterEach(cleanup)

it('fetch and display date', async () => {
  axiosMock.get.mockResolvedValueOnce({
    data: [
      {
        id: 1,
        consultantType: ['gp'],
        appointmentType: ['audio', 'video'],
        time: '2019-11-27T10:11:00.000Z',
      },
      {
        id: 2,
        consultantType: ['specialist', 'gp'],
        appointmentType: ['audio', 'video'],
        time: '2019-12-01T14:16:30.000Z',
      },
    ],
  })
  const { getByTestId, getByText } = render(<App />)
  const resolvedSlot = await waitForElement(() => getByTestId('slot'))
  expect(resolvedSlot).toHaveTextContent('Nov 27, 10:11 amDec 01, 2:16 pm')
  expect(axiosMock.get).toHaveBeenCalledTimes(2)
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
})

it('show the note when it has value nothing when it has not', () => {
  const { getByTestId } = render(<App />)
  const input = getByTestId('symptoms')
  expect(input.value).toBe('')
  fireEvent.change(input, { target: { value: 'cold' } })
  expect(input.value).toBe('cold')
})

it('to be on the page and defined', () => {
  const { getByTestId } = render(<App />)
  const buttonXl = getByTestId('button-xl')
  expect(buttonXl).toBeInTheDocument()
})

it('to have babylone health logo', () => {
  const { getByTestId } = render(<App />)
  const babylonLogo = getByTestId('babylon-logo')
  expect(babylonLogo).toBeInTheDocument()
})

it('match the snapshot', () => {
  const { asFragment } = render(<App />)

  expect(asFragment()).toMatchSnapshot()
})

it('check it has loading', () => {
  const { getByText } = render(<App />)

  expect(getByText('loading')).toBeInTheDocument()
})
