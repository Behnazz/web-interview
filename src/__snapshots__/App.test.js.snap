// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <div
      class="app-header"
    >
      <img
        alt="Babylon Health"
        class="app-logo"
        data-testid="babylon-logo"
        src="logo.png"
      />
    </div>
    <h1
      class="h1"
    >
      New appointment
    </h1>
    <span
      class="app-user"
    >
      <img
        alt="User Avatar"
        class="app-user-logo"
        data-testid="avatar"
      />
      <h3
        class="user-name"
      >
         
      </h3>
    </span>
    <div
      class="section"
    >
      <span
        class="icon"
      >
        <i
          class="fa fa-stethoscope"
        />
        <h2>
          Consultant Type
        </h2>
      </span>
      <div
        class="container"
      >
        <button
          class="button "
        >
          Gp
        </button>
        <button
          class="button "
        >
          Therapist
        </button>
        <button
          class="button "
        >
          Physio
        </button>
        <button
          class="button "
        >
          Specialist
        </button>
      </div>
      <div
        class="section"
      >
        <span
          class="icon"
        >
          <i
            class="fa fa-clock-o"
          />
          <h2>
            Appointment Date and time
          </h2>
        </span>
        <div
          class="container"
          data-testid="slot"
        >
          <h3>
            loading
          </h3>
        </div>
      </div>
      <div
        class="section"
      >
        <span
          class="icon"
        >
          <i
            class="fa fa-video-camera"
          />
          <h2>
            Appointment Type
          </h2>
        </span>
        <div
          class="container"
        >
          <button
            class="button "
          >
            Audio
          </button>
          <button
            class="button "
          >
            Video
          </button>
        </div>
      </div>
      <div
        class="section"
      >
         
        <span
          class="icon"
        >
          <i
            class="fa fa-comments-o"
          />
          <h2>
            Notes
          </h2>
        </span>
        <textarea
          aria-label="note for the symptoms"
          class="note"
          data-testid="symptoms"
          placeholder="Describe your symptoms"
          type="text"
        />
      </div>
      <hr />
      <div>
        <button
          class="button-xl"
          data-testid="button-xl"
        >
          Book
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
